lib/src/clojure_lsp/feature/diagnostics.clj

```diff
@@ -144,42 +143,59 @@
         (when (not= :unknown (shared/uri->file-type uri))
           (sync-lint-file! uri db))))))
 
-(defn ^:private pmap-light
-  "Call claypoole pmap with less threads than pmap to avoid topping cpu."
-  [f coll]
-  (let [threadpool-size (int (Math/ceil (/ (.. Runtime getRuntime availableProcessors) 3)))]
-    (cp/upmap threadpool-size f coll)))
-
 (defn ^:private unused-public-vars-lint!
-  [definitions project-analysis {:keys [config reg-finding!]} max-parallelize? db]
-  (let [parallelize-fn (if max-parallelize? pmap pmap-light)]
-    (->> definitions
-         (remove (partial exclude-public-diagnostic-definition? config))
-         (parallelize-fn #(when (= 0 (count (q/find-references project-analysis % false db)))
-                            %))
-         (remove nil?)
-         (mapv #(reg-unused-public-var-element! % reg-finding! config))
-         (group-by :filename))))
-
-(defn ^:private project-definitions [project-analysis]
-  (concat (q/find-all-var-definitions project-analysis)
-          (q/find-all-keyword-definitions project-analysis)))
-
-(defn ^:private file-definitions [project-analysis filename]
-  (concat (q/find-var-definitions project-analysis filename false)
-          (q/find-keyword-definitions project-analysis filename)))
+  [var-defs kw-defs project-analysis {:keys [config reg-finding!]}]
+  (let [var-definitions (remove (partial exclude-public-diagnostic-definition? config) var-defs)
+        var-nses (set (map :ns var-definitions)) ;; optimization to limit usages to internal namespaces, or in the case of a single file, to its namespaces
+        var-usages (into #{}
+                         (comp
+                           (q/xf-all-var-usages-to-namespaces var-nses)
+                           (map (juxt :to :name)))
+                         project-analysis)
+        var-used? (fn [var-def]
+                    (some (fn [var-name]
+                            (contains? var-usages [(:ns var-def) var-name]))
```

Comment:
I wonder how big var-usages is here, probably huge for some projects, glad to know the performance Is still fast for a huge set which makes sense for a JVM HashSet, quite clever code!

Reply:
Yeah, HashSet lookup should be quite fast, even if it's huge. But actually I doubt it's that big, even for large projects. The biggest files only define a few thousands vars, and the usages set contains (at most) one entry for each definition. And even though project-analysis contains tons of usages of clojure.core, that's all been filtered out by this point.

It's the keyword usages set that gets giant. Since any keyword—even a non-namespaced keyword—can be "registered", the keyword usages set has to include every keyword used anywhere. That's actually the main reason that the vars and keywords are handled separately. If you mix the var and keyword usages together, the var-used? query gets much slower.

Reply:
hum, makes sense! I didn't think keywords would be bigger, but I can see how far that could go, well, glad is not a issue ATM and probably won't be!

Reply:
There's a common lesson here and in the work on completing clojure.core symbols: nesting loops over the analysis is likely to cause performance problems. And there's a common solution: loop once to create an index, then loop again to use the index. Here the index is a set, because all we care about is "which definitions are in the set of used definitions?". In other cases you might need the index to be a hashmap, for example if the question were "what are the usages of these definitions?".

In the database world, this is the difference between a nested loop join and a hash join. The nested loop joins only work well when either A) you have two small datasets, or B) you're worried about using too much memory constructing a hash index of the first dataset. The only other difference between this and the database world is that the databases cache some of their indexes. Computing the indexes at query time is slower than reading them from cache, but under the right conditions, not as slow as a nested loop. For databases that have some sort of EXPLAIN QUERY tool, you can see that sometimes they'll choose to construct an in-memory hash index when they don't already have one cached.

It'd be good to look for other nested loops over the analysis. Here are two: https://github.com/clojure-lsp/clojure-lsp/blob/3bf133a52e358bfa4868ef29e935e0fbfdb076dc/lib/src/clojure_lsp/feature/file_management.clj#L76-L88

Reply:
looks good!
The notify-references looks indeed another candidate to that kind of refactor 👍🏻